app/code/Magento/Quote/Model/Cart/AddProductsToCart.php

```diff
@@ -0,0 +1,222 @@
+        $this->maskedQuoteIdToQuoteId = $maskedQuoteIdToQuoteId;
+        $this->requestBuilder = $requestBuilder;
+    }
+
+    /**
+     * Add cart items to the cart
+     *
+     * @param string $maskedCartId
+     * @param Data\CartItem[] $cartItems
+     * @return AddProductsToCartOutput
+     * @throws NoSuchEntityException Could not find a Cart with provided $maskedCartId
+     */
+    public function execute(string $maskedCartId, array $cartItems): AddProductsToCartOutput
+    {
+        $cartId = $this->maskedQuoteIdToQuoteId->execute($maskedCartId);
+        $cart = $this->cartRepository->get($cartId);
+
+        foreach ($cartItems as $n => $cartItem) {
+            $this->addItemToCart($cart, $cartItem, $n);
+        }
+        if ($cart->getData('has_error')) {
+            $errors = $cart->getErrors();
+
+            /** @var MessageInterface $error */
+            foreach ($errors as $error) {
+                $this->addError($error->getText());
+            }
+        }
+        $this->cartRepository->save($cart);
+
+        return $this->prepareErrorOutput($cart);
+    }
+
+    /**
+     * Adds a particular item to the shopping cart
+     *
+     * @param CartInterface|Quote $cart
+     * @param Data\CartItem $cartItem
+     * @param int $cartItemPosition
+     */
+    private function addItemToCart(CartInterface $cart, Data\CartItem $cartItem, int $cartItemPosition): void
+    {
+        $sku = $cartItem->getSku();
+
+        try {
+            $product = $this->productRepository->get($sku, false, null, true);
+        } catch (NoSuchEntityException $e) {
+            $this->addError(
+                __('Could not find a product with SKU "%sku"', ['sku' => $sku])->render(),
+                $cartItemPosition,
+                self::ERROR_PRODUCT_NOT_FOUND
+            );
+
+            return;
+        }
+
+        try {
+            $result = $cart->addProduct($product, $this->requestBuilder->build($cartItem));
+        } catch (\Throwable $e) {
+            $this->addError(
+                __(
+                    'Could not add the product with SKU %sku to the shopping cart: %message',
+                    ['sku' => $sku, 'message' => $e->getMessage()]
+                )->render(),
+                $cartItemPosition
+            );
+            return;
+        }
+
+        if (is_string($result)) {
+            $errors = array_unique(explode("\n", $result));
+            foreach ($errors as $error) {
+                $this->addError(__($error)->render(), $cartItemPosition);
+            }
+        }
+    }
+
+    /**
+     * Add order line item error
+     *
+     * @param string $message
+     * @param int $cartItemPosition
+     * @param string|null $code
+     * @return void
+     */
+    private function addError(string $message, int $cartItemPosition = 0, string $code = ''): void
+    {
+        $this->errors[] = new Data\Error(
+            $message,
+            $code ?? $this->getErrorCode($message),
```

Comment:
when $code = "", this validation wont work. GraphQl will error out. Please change validation to something like this
empty($code) ? $this->getErrorCode($message) : $code,

You can replicate this issue by trying to get an out of stock error.

Reply:
Good point, I've revised this behavior slightly in the recent commits.